### PR TITLE
Foreman/Satellite el8toel9 upgrade

### PR DESCRIPTION
**Pull request by evgeni**
* move common Satellite upgrade code to `common` so it can be used by `7to8` and `8to9` upgrades
* add `8to9` specific code to allow Foreman/Satellite upgrades from EL8 to EL9
* refresh collation version if pulp-ansible (and its custom collation) is present
* refactor satellite_upgrade_services to use SystemdServicesTasks

### CODE REVIEW

File: repos/system_upgrade/common/models/satellite.py
Comment (evgeni) [inline repos/system_upgrade/common/models/satellite.py:26]: Is it OK to rename the field in the model, when I know we're the only consumer?
We could also add a new field (or keep the old, not-to-accurate, name).

File: repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py
Comment (evgeni) [inline repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py:14]: this is just moved code from `satellite_upgrade_data_migration/actor.py` as those steps are el-version-agnostic, while the code remaining in `satellite_upgrade_data_migration/actor.py` is only needed for 7to8 due to the SCL usage

File: repos/system_upgrade/el8toel9/actors/satellite_upgrade_facts/actor.py
Comment (evgeni) [inline repos/system_upgrade/el8toel9/actors/satellite_upgrade_facts/actor.py:13]: obviously will need to be updated to the right Satellite version, once we know what it will be.

File: repos/system_upgrade/common/models/satellite.py
Comment (pirat89) [reply repos/system_upgrade/common/models/satellite.py:26]: no, you should  go with the deprecation process, which means 6month protection that the original field will exists and will have still the same functionality as before. with the point that upstream documentation is updated to mention this one field became deprecated. docstring should be updated as well.

Comment (evgeni) [reply repos/system_upgrade/common/models/satellite.py:26]: I guess I'll just stick with the old name then ;)

File: repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py
Comment (fernflower) [inline repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py:34]: Do we really need silencing the linter here? I'd say `FileNotFoundError` and `OSError` should cover everything

**Review by fernflower — COMMENTED** [review]
I see that this is not a ready to go pr (at least given the fact SATELLITE_VERSIONs are not set in stone yet). I will move it to draft so that no one on a merging spree gets the wrong idea :)

File: repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py
Comment (evgeni) [reply repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py:34]: Nah, it was just blindly copied from the old code.

Given `FileNotFoundError` is an `OSError`, I think the latter is sufficient?

**Comment by evgeni** [comment]
> I see that this is not a ready to go pr (at least given the fact SATELLITE_VERSIONs are not set in stone yet). I will move it to draft so that no one on a merging spree gets the wrong idea :)

While not ready for Satellite consumption, this can today upgrade upstream Foreman just fine, so I'd be interested to get it in as-is, if you'd be OK with that.

File: repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py
Comment (fernflower) [reply repos/system_upgrade/common/actors/satellite_upgrade_services/actor.py:34]: Sure, OSError should be enough if you don't care about the reason

**Comment by fernflower** [comment]
Got it, removing draft :)

File: repos/system_upgrade/el8toel9/actors/satellite_upgrade_facts/actor.py
Comment (fernflower) [reply repos/system_upgrade/el8toel9/actors/satellite_upgrade_facts/actor.py:13]: Could you please add a TODO here then? If this can get in before the right version is there?

Comment (evgeni) [reply repos/system_upgrade/el8toel9/actors/satellite_upgrade_facts/actor.py:13]: Actually, we can do better, we can compute the correct name for the el9 repo based on the list of enabled el8 repos!
Will do on Monday.
(Ideas you have when doing dishes and listening to 90ies pop music 😅)

File: .pylintrc
Comment (evgeni) [inline .pylintrc:60]: probably should be an own PR with nicer explanation, but I just wanted to silence it here for now

Comment (abadger) [reply .pylintrc:60]: The comment could be

> Yield from cannot be used until we require python3.3 or greater

(I'm not sure we'd want to use yield from everywhere possible then but ot will do for now)

Comment (evgeni) [reply .pylintrc]: Thanks, updated and opened https://github.com/oamg/leapp-repository/pull/1187 with this commit only

**Review by fernflower — APPROVED** [review]
lgtm codewise.
But looks like the final decision on mergeability is on @mkluson and/or after we get some proof that is works from the Satellite team.

**Comment by mkluson** [comment]
> lgtm codewise. But looks like the final decision on mergeability is on @mkluson and/or after we get some proof that this works as expected from the Satellite team.

Ack, I am in touch with Satellite team and I will coordinate the merge. Thanks

**Review by pirat89 — CHANGES_REQUESTED** [review]
seems there are some issues about the solution based on the internal communication. so I am setting required changes to prevent accidental merge. Feel free to drop it once all problems are resolved and the solution is confirmed to be working as expected.

@evgeni  fill please also the PR description.

**Comment by evgeni** [comment]
@pirat89 FWIW, upgrades work when `foreman-protector` is disabled (as it is in upstream Foreman, or can be done manually in Satellite), so I wouldn't necessarily block this PR on that. Nevertheless, as discussed, I have a patch against `foreman-protector` to self-disable when it detects a Leapp upgrade: https://github.com/theforeman/foreman_maintain/pull/828

Description updated.

**Comment by evgeni** [comment]
@pirat89 you asked me to re-check with the latest changes in `master` and things look good on my end :)

**Comment by pirat89** [comment]
> @pirat89 you asked me to re-check with the latest changes in `master` and things look good on my end :)

Thanks for the check.

**Comment by pirat89** [comment]
@evgeni ha! one more rebase needed due to major change in tests! Introduced by #1209

**Comment by evgeni** [comment]
No worries, rebases are cheap :)

**Comment by lpramuk** [comment]
I have successfully tested this PR for both Satellite and Capsule using:
`leapp-0.17.0-100.20240430083919859552.master.3.gb21450a.el8`
`leapp-repository-0.20.0-0.20240426090422205877.pr1181.30.gc00d93a3.el8`

**Review by fernflower — DISMISSED** [review]
lgtm

**Comment by fernflower** [comment]
@evgeni Could you please squash commits into logical blocks (should not be one commit) and add description to each of those squashed commits' messages?
Happy to merge it afterwards.

**Comment by evgeni** [comment]
Will do.
Also just realized new PES data was updated, so I'll check if we still need the workaround for RHEL-32759 (my expectation is not)

**Comment by evgeni** [comment]
It's not needed \o/

**Comment by evgeni** [comment]
Thanks everyone!

**Comment by evgeni** [comment]
1. Yes, utils is not for Satellite/Capsule systems. It's for systems that need Satellite cli tools (hammer etc), but are not Satellites.
2. This actor is only for Satellite/Capsule systems. Clients don't need special handling (besides repo-enablement, and that should happen via PES)
3. Sat on EL8 uses PostgreSQL 13, Sat on EL9 uses PostgreSQL 13. So we only need to run `reindexdb` to ensure consistent indexes, no PostgreSQL-upgrade is required.

**Pull request merged**
